Enable reading of FileStatistics in presto-orc

sdruzkin opened the pull request:
Move intermediate encryption keys from the OrcRecordReader to the OrcReader's constructor.
Enable reading of FileStatistics for both encrypted and unencrypted ORC files.

== Test plan ==
```
mvn clean install -DskipTests
mvn clean install -pl presto-orc -pl presto-hive -pl presto-raptor
```

== RELEASE NOTES ==

General Changes
* Move IEK from OrcRecordReader to OrcReader's constructor
* Enable reading of FileStatistics for ORC files

yingsu00 commented:
> This is the reason why I had to modify the validation process in the OrcWriteValidation.
Hi @sdruzkin Will you please add reference to the PR or commit ids that you "modified the validation process in the OrcWriteValidation"?

yingsu00 reviewed (commented):
@sdruzkin Sergii, I might have misunderstood your commit message that you referred to some previous commit. If you didn't mean that but just explain why you made the change in THIS commit, I think it's better to use imperative mode instead of "I had to...". 

I ran out of time reading the PR. I will continue tomorrow.

rschlussel commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
     }
 
+    @Override
+    public List<ColumnStatistics> decryptFileStatistics(HiveWriterVersion hiveWriterVersion,
```
This is doing more than decryption.  I would call this decryptAndCombineFileStatistics or something.

rschlussel reviewed (commented):
Doing the decryption in the RecordReader is sort of ugly, though I see why you had to do it. It would be nice, though, to pass the encryption information to the OrcReader constructor instead and be able to do the decryption as part of reading the footer.  The tricky thing is that we use some info from the reader  to construct the key map(e.g the types in DwrfEncryptionMetadata.toKeyMap, and the column names in HiveUitl.getPhysicalHiveColumnHandles()).  Maybe there's a way to pass  a map that contains hive column Index + name + subfield mapped to the key into the orcreader, and then in the reader have the logic to map that to the correct Orc column.

sujay-jain commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
+        checkState(encryptionLibrary.isPresent(), "encryption library is not set");
+        ColumnStatistics[] decryptedFileStats = fileStats.toArray(new ColumnStatistics[0]);
+
+        List<byte[]> stripeKeys = fileStripes.get(0).getKeyMetadata();
```
Is ```fileStripes``` guaranteed to be non-empty?

sujay-jain commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
+        for (int groupIdx = 0; groupIdx < encryptionGroups.size(); groupIdx++) {
+            EncryptionGroup encryptionGroup = encryptionGroups.get(groupIdx);
+            DwrfDataEncryptor decryptor = null;
+            for (int i = 0; i < encryptionGroup.getNodes().size(); i++) {
```
let's extract ```encryptionGroup.getNodes()``` to a variable so we don't call it twice

sujay-jain commented on presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfMetadataReader.java:
```diff
@@ -112,6 +118,84 @@ public Footer readFooter(HiveWriterVersion hiveWriterVersion, InputStream inputS
+                    // The key in the footer takes priority over the key in the first stripe.
+                    byte[] encryptedDataKeyWithMeta;
+                    if (!encryptionGroup.getKeyMetadata().isPresent()) {
+                        encryptedDataKeyWithMeta = encryptionGroup.getKeyMetadata().get().byteArray();
```
if ```!encryptionGroup.getKeyMetadata().isPresent()``` is true then ```encryptionGroup.getKeyMetadata().get()``` will blow up?

yingsu00 commented on presto-orc/src/test/java/com/facebook/presto/orc/TestingOrcPredicate.java:
```diff
@@ -83,57 +83,57 @@ public static OrcPredicate createOrcPredicate(int columnIndex, Type type, Iterab
                             .map(value -> value == null ? null : ((Number) value).longValue())
                             .collect(toList()),
-                    format == DWRF);
+                    false);
```
`createOrcPredicate` is used by `OrcTester` and is used by all block tests. Is there any reason you set the `noFileStats` to false for all the tests? The tests could be on other formats like ORC, are they all without file stats?

sdruzkin commented:
> > This is the reason why I had to modify the validation process in the OrcWriteValidation.
> 
> Hi @sdruzkin Will you please add reference to the PR or commit ids that you "modified the validation process in the OrcWriteValidation"?

Done

> @sdruzkin Sergii, I might have misunderstood your commit message that you referred to some previous commit. If you didn't mean that but just explain why you made the change in THIS commit, I think it's better to use imperative mode instead of "I had to...".

You are right, the more appropriate place for this note is the comment in the PR. I removed these lines from the commit message and PR description.

> Doing the decryption in the RecordReader is sort of ugly,
I will look at this tomorrow.

sdruzkin commented:
I moved the encryption keys from the OrcRecordReader to the OrcReader in a first commit per @rschlussel's request. Then I added loading and decryption of file stats in the second commit to have clear separation between these two changes. 

I also removed changes to the validation process because now the writer validation has full access to the decrypted stats in the OrcReader.

mbasmanova commented:
@sdruzkin Sergii, I see "This branch has conflicts that must be resolved". Would you rebase on top of latest master and update the PR. @rschlussel Rebecca, is this good to go otherwise?

rschlussel reviewed (approved):
Looks good.  Really like the new key provider abstraction.
